op-chain-ops/cmd/check-canyon/main.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb/memorydb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+func PreCanyonEncode(receipts types.Receipts) [][]byte {
+	for _, receipt := range receipts {
+		if receipt.Type == types.DepositTxType {
+			receipt.DepositReceiptVersion = nil
+		}
+	}
+	var out [][]byte
+	for i := range receipts {
+		var buf bytes.Buffer
+		receipts.EncodeIndex(i, &buf)
+		out = append(out, buf.Bytes())
+	}
+	return out
+}
+
+func PostCanyonEncode(receipts types.Receipts) [][]byte {
+	v := uint64(1)
+	for _, receipt := range receipts {
+		if receipt.Type == types.DepositTxType {
+			receipt.DepositReceiptVersion = &v
+		}
+	}
+	var out [][]byte
+	for i := range receipts {
+		var buf bytes.Buffer
+		receipts.EncodeIndex(i, &buf)
+		out = append(out, buf.Bytes())
+	}
+	return out
+}
+
+func HashList(list [][]byte) common.Hash {
+	t := trie.NewEmpty(trie.NewDatabase(rawdb.NewDatabase(memorydb.New()), nil))
+	for i, value := range list {
+		var index []byte
+		val := make([]byte, len(value))
+		copy(val, value)
+		index = rlp.AppendUint64(index, uint64(i))
+		if err := t.Update(index, val); err != nil {
+			panic(err)
+		}
+	}
+	return t.Hash()
+}
```

Comment:
```suggestion
type rawReceipts [][]byte

func (s rawReceipts) Len() int { return len(s) }
func (s rawReceipts) EncodeIndex(i int, w *bytes.Buffer) {
	w.Write(s[i])
}

func HashList(list [][]byte) common.Hash {
	hasher := trie.NewStackTrie(nil)
	return types.DeriveSha(rawReceipts(list), hasher)
}
```

Reply:
Oh, nice cleanup. The manual trie was from when I was trying to do MPT proofs against single encoded receipts